src/devices/GoPiGo3/GoPiGo3.cs

```diff
@@ -0,0 +1,900 @@
+        {
+            int retVal = -1;
+            byte[] outArray = { SpiAddress, (byte)MessageType, 0, 0, 0, 0, };
+            byte[] reply = SpiTransferArray(outArray);
+
+            if (reply[3] == 0xA5)
+            {
+                retVal = (int)(reply[4] << 8) | reply[5];
+            }
+            else
+            {
+                throw new IOException($"{nameof(SpiRead16)} : no SPI response");
+            }
+
+            return retVal;
+        }
+
+        /// <summary>
+        /// Read a 8 bit value over SPI
+        /// </summary>
+        /// <param name="MessageType">The SPI message type</param>
+        /// <returns>Return the vallue read</returns>
+        public int SpiRead8(SpiMessageType MessageType)
+        {
+            byte[] outArray = { SpiAddress, (byte)MessageType, 0, 0, 0 };
+            byte[] reply = SpiTransferArray(outArray);
+            Console.WriteLine($"{reply[0]} {reply[1]} {reply[2]} {reply[3]} {reply[4]}");
+            return (reply[3] == 0xA5) ? reply[4] : throw new IOException($"{nameof(SpiRead16)} : no SPI response");
+        }
+
+        /// <summary>
+        /// Send a 8 bit value over SPI
+        /// </summary>
+        /// <param name="MessageType">The SPI message type</param>
+        /// <param name="Value">The value to be sent</param>
+        public void SpiWrite8(SpiMessageType MessageType, int Value)
+        {
+            byte[] outArray = { SpiAddress, (byte)MessageType, (byte)(Value & 0xFF) };
+            SpiTransferArray(outArray);
+        }
+
+        /// <summary>
+        /// Send a 16 bit value over SPI
+        /// </summary>
+        /// <param name="MessageType">The SPI message type</param>
+        /// <param name="Value">The value to be sent</param>
+        public void SpiWrite16(SpiMessageType MessageType, int Value)
+        {
+            byte[] outArray = { SpiAddress, (byte)MessageType, (byte)((Value >> 8) & 0xFF), (byte)(Value & 0xFF) };
+            SpiTransferArray(outArray);
+        }
+
+        /// <summary>
+        /// Send a 24 bit value over SPI
+        /// </summary>
+        /// <param name="MessageType">The SPI message type</param>
+        /// <param name="Value">The value to be sent</param>
+        public void SpiWrite24(SpiMessageType MessageType, int Value)
+        {
+            byte[] outArray = { SpiAddress, (byte)MessageType, (byte)((Value >> 16) & 0xFF), (byte)((Value >> 8) & 0xFF), (byte)(Value & 0xFF) };
+            SpiTransferArray(outArray);
+        }
+
+        /// <summary>
+        /// Send a 32 bit value over SPI
+        /// </summary>
+        /// <param name="MessageType">The SPI message type</param>
+        /// <param name="Value">The value to be sent</param>
+        public void SpiWrite32(SpiMessageType MessageType, int Value)
+        {
+            byte[] outArray = { SpiAddress, (byte)MessageType, (byte)((Value >> 24) & 0xFF), (byte)((Value >> 16) & 0xFF), (byte)((Value >> 8) & 0xFF), (byte)(Value & 0xFF) };
```

Comment:
I'd take uint as an input here, signed int does not do what you might expect when you shift bits

Reply:
it's working the same way in this case. And as it's working correctly even for negative numbers, I would leave it like this

Reply:
fine with leaving, but FYI:

Here is the output:

```
x=11111111111111111111111111111111    y=11111111111111111111111111111111

shifting right (int)
11111111111111111111111111111111
11111111111111111111111111111111
11111111111111111111111111111111
11111111111111111111111111111111
11111111111111111111111111111111
11111111111111111111111111111111
11111111111111111111111111111111
11111111111111111111111111111111
11111111111111111111111111111111
11111111111111111111111111111111
11111111111111111111111111111111
11111111111111111111111111111111
11111111111111111111111111111111
11111111111111111111111111111111
11111111111111111111111111111111
11111111111111111111111111111111
11111111111111111111111111111111
11111111111111111111111111111111
11111111111111111111111111111111
11111111111111111111111111111111
11111111111111111111111111111111
11111111111111111111111111111111
11111111111111111111111111111111
11111111111111111111111111111111
11111111111111111111111111111111
11111111111111111111111111111111
11111111111111111111111111111111
11111111111111111111111111111111
11111111111111111111111111111111
11111111111111111111111111111111
11111111111111111111111111111111
11111111111111111111111111111111

shifting left (int)
11111111111111111111111111111111
11111111111111111111111111111110
11111111111111111111111111111100
11111111111111111111111111111000
11111111111111111111111111110000
11111111111111111111111111100000
11111111111111111111111111000000
11111111111111111111111110000000
11111111111111111111111100000000
11111111111111111111111000000000
11111111111111111111110000000000
11111111111111111111100000000000
11111111111111111111000000000000
11111111111111111110000000000000
11111111111111111100000000000000
11111111111111111000000000000000
11111111111111110000000000000000
11111111111111100000000000000000
11111111111111000000000000000000
11111111111110000000000000000000
11111111111100000000000000000000
11111111111000000000000000000000
11111111110000000000000000000000
11111111100000000000000000000000
11111111000000000000000000000000
11111110000000000000000000000000
11111100000000000000000000000000
11111000000000000000000000000000
11110000000000000000000000000000
11100000000000000000000000000000
11000000000000000000000000000000
10000000000000000000000000000000

shifting right (uint)
11111111111111111111111111111111
01111111111111111111111111111111
00111111111111111111111111111111
00011111111111111111111111111111
00001111111111111111111111111111
00000111111111111111111111111111
00000011111111111111111111111111
00000001111111111111111111111111
00000000111111111111111111111111
00000000011111111111111111111111
00000000001111111111111111111111
00000000000111111111111111111111
00000000000011111111111111111111
00000000000001111111111111111111
00000000000000111111111111111111
00000000000000011111111111111111
00000000000000001111111111111111
00000000000000000111111111111111
00000000000000000011111111111111
00000000000000000001111111111111
00000000000000000000111111111111
00000000000000000000011111111111
00000000000000000000001111111111
00000000000000000000000111111111
00000000000000000000000011111111
00000000000000000000000001111111
00000000000000000000000000111111
00000000000000000000000000011111
00000000000000000000000000001111
00000000000000000000000000000111
00000000000000000000000000000011
00000000000000000000000000000001

shifting left (uint)
11111111111111111111111111111111
11111111111111111111111111111110
11111111111111111111111111111100
11111111111111111111111111111000
11111111111111111111111111110000
11111111111111111111111111100000
11111111111111111111111111000000
11111111111111111111111110000000
11111111111111111111111100000000
11111111111111111111111000000000
11111111111111111111110000000000
11111111111111111111100000000000
11111111111111111111000000000000
11111111111111111110000000000000
11111111111111111100000000000000
11111111111111111000000000000000
11111111111111110000000000000000
11111111111111100000000000000000
11111111111111000000000000000000
11111111111110000000000000000000
11111111111100000000000000000000
11111111111000000000000000000000
11111111110000000000000000000000
11111111100000000000000000000000
11111111000000000000000000000000
11111110000000000000000000000000
11111100000000000000000000000000
11111000000000000000000000000000
11110000000000000000000000000000
11100000000000000000000000000000
11000000000000000000000000000000
10000000000000000000000000000000
```

```csharp
using System;

public class Program
{
  public static string ToBinary32(int x)
  {
    return Convert.ToString(x, 2).PadLeft(32, '0');
  }

  public static string ToBinary32(uint x)
  {
    return Convert.ToString(x, 2).PadLeft(32, '0');
  }

  public static void Main()
  {
    const int x = -1;
    const uint y = unchecked((uint)x);
    Console.WriteLine($"x={ToBinary32(x)}    y={ToBinary32(y)}");

    //void testFunc(string title, Func<)

    Console.WriteLine();
    Console.WriteLine("shifting right (int)");
    for (int i = 0; i < 32; i++)
    {
      Console.WriteLine(ToBinary32(x >> i));
    }

    Console.WriteLine();
    Console.WriteLine("shifting left (int)");
    for (int i = 0; i < 32; i++)
    {
        Console.WriteLine(ToBinary32(x << i));
    }


    Console.WriteLine();
    Console.WriteLine("shifting right (uint)");
    for (int i = 0; i < 32; i++)
    {
        Console.WriteLine(ToBinary32(y >> i));
    }

    Console.WriteLine();
    Console.WriteLine("shifting left (uint)");
    for (int i = 0; i < 32; i++)
    {
        Console.WriteLine(ToBinary32(y << i));
    }
  }
}
```